src/worker/activities/local_activities.rs

```diff
@@ -211,19 +288,42 @@ impl LocalActivityManager {
             NewOrRetry::Retry { in_flight, attempt } => (in_flight, attempt),
         };
         let orig = new_la.clone();
+        let id = ExecutingLAId {
+            run_id: new_la.workflow_exec_info.run_id.clone(),
+            seq_num: new_la.schedule_cmd.seq,
+        };
         let sa = new_la.schedule_cmd;
 
         let mut dat = self.dat.lock();
         // If this request originated from a local backoff task, clear the entry for it. We
         // don't await the handle because we know it must already be done, and there's no
         // meaningful value.
-        dat.backing_off_tasks.remove(&ExecutingLAId {
-            run_id: new_la.workflow_exec_info.run_id.clone(),
-            seq_num: sa.seq,
-        });
+        dat.backing_off_tasks.remove(&id);
+
+        // If this task sat in the queue for too long, return a timeout for it instead
+        if let Some(s2s) = sa.schedule_to_start_timeout.as_ref() {
```

Comment:
Why is this only done as a response to polling?
Notifying the workflow of a timeout shouldn't be coupled with activity polling.

Reply:
Yeah, that's fair. We can perhaps decouple this later as discussed - for now it's not really an issue since langs shouldn't stop polling core and core handles open slots for non-cancel/timeout stuff.